packages/node-core/src/indexer/dictionary/dictionary.service.ts

```diff
@@ -0,0 +1,157 @@
+// Copyright 2020-2023 SubQuery Pte Ltd authors & contributors
+// SPDX-License-Identifier: GPL-3.0
+
+import {EventEmitter2} from '@nestjs/event-emitter';
+import {NETWORK_FAMILY} from '@subql/common';
+import fetch from 'cross-fetch';
+import {NodeConfig} from '../../configure';
+import {getLogger} from '../../logger';
+import {BlockHeightMap} from '../../utils/blockHeightMap';
+import {DictionaryResponse, DictionaryVersion, IDictionary, IDictionaryCtrl} from './types';
+import {subqlFilterBlocksCapabilities} from './utils';
+import {DictionaryV1} from './v1';
+import {DictionaryV2} from './v2';
+import {DictionaryV2Metadata} from './';
+
+async function inspectDictionaryVersion(endpoint: string): Promise<DictionaryVersion> {
+  if (endpoint.includes('/rpc')) {
+    let resp: DictionaryV2Metadata;
+    try {
+      resp = await subqlFilterBlocksCapabilities(endpoint);
+      if (resp.supported.includes('complete')) {
+        return DictionaryVersion.v2Complete;
+      } else {
+        return DictionaryVersion.v2Basic;
+      }
+    } catch (e) {
+      logger.warn(`${e}. Try to use dictionary V1`);
+      return DictionaryVersion.v1;
+    }
+  }
+  return DictionaryVersion.v1;
+}
+
+const logger = getLogger('DictionaryService');
+export abstract class DictionaryService<RFB, FB, DS, D extends DictionaryV1<DS>>
+  implements IDictionaryCtrl<DS, FB, IDictionary<DS, FB>>
+{
+  protected _dictionaries: (DictionaryV2<RFB, FB, DS> | D)[] = [];
+  protected _currentDictionaryIndex: number | undefined;
+  // Keep in memory in order one of dictionary failed then we can switch
+  protected _dictionaryV1Endpoints: string[] = [];
+  protected _dictionaryV2Endpoints: string[] = [];
+
+  constructor(
+    protected chainId: string,
+    protected readonly nodeConfig: NodeConfig,
+    protected readonly eventEmitter: EventEmitter2
+  ) {}
+
+  protected abstract initDictionariesV1(): Promise<D[]>;
+  protected abstract initDictionariesV2(): Promise<DictionaryV2<RFB, FB, DS>[]> | DictionaryV2<RFB, FB, DS>[];
+
+  async initDictionaries(apiGenesisHash: string): Promise<void> {
+    // For now, treat dictionary resolver as V1
+    if (this.nodeConfig.networkDictionaries) {
+      for (const endpoint of this.nodeConfig.networkDictionaries) {
+        const version = await inspectDictionaryVersion(endpoint);
+        if (version === DictionaryVersion.v1) {
+          this._dictionaryV1Endpoints.push(endpoint);
+        } else {
+          this._dictionaryV2Endpoints.push(endpoint);
+        }
+      }
+    }
+    this._dictionaries.push(...(await this.initDictionariesV1()));
+    this._dictionaries.push(...(await this.initDictionariesV2()));
+  }
```

Comment:
I don't think this class would have any distinction between v1 or v2. The init function can take an array of IDictionary. 

That way we could implement a dictionary that isn't v1 or v2 and is specific to a chain. E.g cosmos rpc that can filter by events/transactions, we should be able to implement that without any changes to node core.